src/image/release.py

```diff
@@ -0,0 +1,231 @@
+#!/usr/bin/env python3
+
+"""
+Takes a releases trigger file and created a mapping of all the
+OCI tags that are marked to be release.
+"""
+
+import argparse
+import json
+import os
+import re
+import subprocess
+import yaml
+
+from collections import defaultdict
+from utils.schema.triggers import ImageSchema, KNOWN_RISKS_ORDERED
+
+
+class BadChannel(Exception):
+    """Error validating release channel."""
+
+
+parser = argparse.ArgumentParser()
+parser.add_argument(
+    "--image-trigger",
+    help="Path to the image trigger file.",
+    required=True,
+)
+parser.add_argument(
+    "--image-name",
+    help="Image name. Will infer from --image-trigger if not provided.",
+    required=False,
+    default=None,
+)
+parser.add_argument(
+    "--all-releases",
+    help="Path to the _releases.yaml file.",
+    required=True,
+)
+parser.add_argument(
+    "--all-revision-tags",
+    help="File w/ comma-separated list of all revision (<track>_<rev>) tags.",
+    required=True,
+)
+parser.add_argument(
+    "--ghcr-repo",
+    help="GHCR repo where the image was originally uploaded.",
+    required=True,
+)
+
+args = parser.parse_args()
+img_name = (
+    args.image_name
+    if args.image_name
+    else os.path.abspath(args.image_trigger).split("/")[-2]
+)
+
+print(f"Preparing to release revision tags for {img_name}")
+with open(args.all_revision_tags) as rev_tags_f:
+    all_revision_tags = (
+        rev_tags_f.read().strip().rstrip(",").lstrip(",").split(",")
+    )
+revision_to_track = {}
+for track_revision in all_revision_tags:
+    track, revision = track_revision.split("_")
+    if revision in revision_to_track:
+        msg = (
+            "Each revision can only have 1 canonical tag, "
+            f"but revision {revision} is associated with tracks "
+            f"{track} and {revision_to_track['revision']}!"
+        )
+        raise BadChannel(msg)
+
+    revision_to_track[int(revision)] = track
+
+print(
+    "Revision (aka 'canonical') tags grouped by revision:\n"
+    f"{json.dumps(revision_to_track, indent=2)}"
+)
+
+print(f"Reading all previous releases from {args.all_releases}...")
+tag_mapping_from_all_releases = {}
+try:
+    with open(args.all_releases) as all_releases_fd:
+        all_releases = yaml.safe_load(all_releases_fd).get("channels", {})
+
+    # map the existing tags into a struct similar to tag_mapping_from_trigger
+    for track, risks in all_releases.items():
+        for risk, values in risks.items():
+            if risk not in KNOWN_RISKS_ORDERED:
```

Comment:
it's better to do if risk in  KNOWN_RISKS_ORDERED : tag ... no ?

Reply:
sure why not